File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Star.java
package com.bombbird.terminalcontrol.entities.sidstar;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.airports.Airport;
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;

public class Star extends SidStar {
    private Array<Array<String>> inbound;
    private HashMap<String, Array<Waypoint>> rwyWpts;
    private HashMap<String, Array<int[]>> rwyRestrictions;
    private HashMap<String, Array<Boolean>> rwyFlyOver;

    public Star(Airport airport, JSONObject jo) {
        super(airport, jo);
    }

    public Star(Airport airport, JSONArray wpts, JSONArray restrictions, JSONArray fo, String name) {
        super(airport, wpts, restrictions, fo, name);

        inbound = new Array<>();
        rwyWpts = new HashMap<>();
        rwyRestrictions = new HashMap<>();
        rwyFlyOver = new HashMap<>();

        Array<String> inbound1 = new Array<>();
        inbound1.add("HDG 360");
        inbound.add(inbound1);
    }

    @Override
    public void parseInfo(JSONObject jo) {
        super.parseInfo(jo);

        inbound = new Array<>();
        rwyWpts = new HashMap<>();
        rwyRestrictions = new HashMap<>();
        rwyFlyOver = new HashMap<>();

        JSONObject rwys = jo.getJSONObject("rwys");
        for (String rwy: rwys.keySet()) {
            getRunways().add(rwy);
            Array<Waypoint> wpts = new Array<>();
            Array<int[]> restrictions = new Array<>();
            Array<Boolean> flyOver = new Array<>();

            JSONArray rwyObject = rwys.getJSONArray(rwy);
            for (int i = 0; i < rwyObject.length(); i++) {
                String[] data = rwyObject.getString(i).split(" ");
                String wptName = data[0];
                wpts.add(TerminalControl.radarScreen.waypoints.get(wptName));
                restrictions.add(new int[] {Integer.parseInt(data[1]), Integer.parseInt(data[2]), Integer.parseInt(data[3])});
                boolean fo = data.length > 4 && data[4].equals("FO");
                flyOver.add(fo);
                if (fo) Waypoint.flyOverPts.put(wptName, true);
            }
            rwyWpts.put(rwy, wpts);
            rwyRestrictions.put(rwy, restrictions);
            rwyFlyOver.put(rwy, flyOver);
        }

        JSONArray inbounds = jo.getJSONArray("inbound");
        for (int i = 0; i < inbounds.length(); i++) {
            JSONArray trans = inbounds.getJSONArray(i);
            Array<String> transData = new Array<>();
            for (int j = 0; j < trans.length(); j++) {
                transData.add(trans.getString(j));
            }
            inbound.add(transData);
        }
    }

    public Array<String> getAllInboundWpt() {
        Array<String> wpts = new Array<>();
        for (int i = 0; i < inbound.size; i++) {
            Array<String> inboundPts = inbound.get(i);
            if (inboundPts.size > 1) {
                wpts.add(inboundPts.get(1).split(" ")[1]);
            }
        }
        wpts.add(getWaypoints().get(0).getName());

        return wpts;
    }

    public Array<String> getRandomInbound() {
        return inbound.get(MathUtils.random(inbound.size - 1));
    }

    public Array<Waypoint> getRwyWpts(String runway) {
        return rwyWpts.get(runway);
    }

    public Array<int[]> getRwyRestrictions(String runway) {
        return rwyRestrictions.get(runway);
    }

    public Array<Boolean> getRwyFlyOver(String runway) {
        return rwyFlyOver.get(runway);
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Star.kt
package com.bombbird.terminalcontrol.entities.sidstar

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

class Star : SidStar {
    private lateinit var inbound: Array<Array<String>>
    private lateinit var rwyWpts: HashMap<String, Array<Waypoint>>
    private lateinit var rwyRestrictions: HashMap<String, Array<IntArray>>
    private lateinit var rwyFlyOver: HashMap<String, Array<Boolean>>

    constructor(airport: Airport, jo: JSONObject) : super(airport) {
        parseInfo(jo)
    }

    constructor(airport: Airport, wpts: JSONArray, restrictions: JSONArray, fo: JSONArray, name: String) : super(airport, wpts, restrictions, fo, name) {
        inbound = Array()
        rwyWpts = HashMap()
        rwyRestrictions = HashMap()
        rwyFlyOver = HashMap()
        val inbound1 = Array<String>()
        inbound1.add("HDG 360")
        inbound.add(inbound1)
    }

    override fun parseInfo(jo: JSONObject) {
        super.parseInfo(jo)
        inbound = Array()
        rwyWpts = HashMap()
        rwyRestrictions = HashMap()
        rwyFlyOver = HashMap()
        val rwys = jo.getJSONObject("rwys")
        for (rwy in rwys.keySet()) {
            runways.add(rwy)
            val wpts = Array<Waypoint>()
            val restrictions = Array<IntArray>()
            val flyOver = Array<Boolean>()
            val rwyObject = rwys.getJSONArray(rwy)
            for (i in 0 until rwyObject.length()) {
                val data = rwyObject.getString(i).split(" ".toRegex()).toTypedArray()
                val wptName = data[0]
                wpts.add(radarScreen.waypoints[wptName])
                restrictions.add(intArrayOf(data[1].toInt(), data[2].toInt(), data[3].toInt()))
                val fo = data.size > 4 && data[4] == "FO"
                flyOver.add(fo)
            }
            rwyWpts[rwy] = wpts
            rwyRestrictions[rwy] = restrictions
            rwyFlyOver[rwy] = flyOver
        }
        val inbounds = jo.getJSONArray("inbound")
        for (i in 0 until inbounds.length()) {
            val trans = inbounds.getJSONArray(i)
            val transData = Array<String>()
            for (j in 0 until trans.length()) {
                transData.add(trans.getString(j))
            }
            inbound.add(transData)
        }
    }

    val allInboundWpt: Array<String>
        get() {
            val wpts = Array<String>()
            for (i in 0 until inbound.size) {
                val inboundPts = inbound[i]
                if (inboundPts.size > 1) {
                    wpts.add(inboundPts[1].split(" ".toRegex()).toTypedArray()[1])
                }
            }
            wpts.add(routeData.waypoints.get(0).name)
            return wpts
        }

    val randomInbound: Array<String>
        get() = inbound[MathUtils.random(inbound.size - 1)]

    fun getRwyWpts(runway: String): Array<Waypoint> {
        return rwyWpts[runway] ?: Array()
    }

    fun getRwyRestrictions(runway: String): Array<IntArray> {
        return rwyRestrictions[runway] ?: Array()
    }

    fun getRwyFlyOver(runway: String): Array<Boolean> {
        return rwyFlyOver[runway] ?: Array()
    }
}